### PR TITLE
!!! FEATURE: References on creation and Copy

### PR DESCRIPTION
**Pull request by kitsunet**
This reworks references so that multiple reference properties can be set via a single command and also references can be attached to `CreateNodeAggregateWithNode` which is also used for copying nodes.

The serialization format is adapted to allow multiple reference properties, which also affects all behat tests with references.

To migrate existing events run:

`./flow migrateevents:migratesetreferencestomultinameformat`

Marked breaking as it requires an event migration.

This requires https://github.com/neos/neos-ui/pull/3844 for the UI

A behat test to show reference copies work was added.

**Upgrade Instructions**

The API to set references changed a little, please adjust accordingly.
This is a simple example which covers all cases:

```php
SetNodeReferences::create(
    $node->workspaceName,
    $node->aggregateId,
    $node->originDimensionSpacePoint,
    NodeReferencesToWrite::create(
        // sets the 3 nodes as references
        NodeReferencesForName::fromTargets(
            ReferenceName::fromString('first-reference'),
            NodeAggregateIds::fromArray(['node-aggregate-id-1', 'node-aggregate-id-2', 'node-aggregate-id-3'])
        ),
        // unset previously set references (as we do NOT merge the values)
        NodeReferencesForName::createEmpty(
            ReferenceName::fromString('second-reference')
        ),
        // create a simple node reference and additionally a reference with an additional property
        NodeReferencesForName::fromReferences(
            ReferenceName::fromString('third-reference'),
            [
                NodeReferenceToWrite::fromTarget(NodeAggregateId::fromString('node-aggregate-id-6')),
                NodeReferenceToWrite::fromTargetAndProperties(NodeAggregateId::fromString('node-aggregate-id-5'), PropertyValuesToWrite::fromArray([
                    'propertyOnReference' => 'Hi im living on the edge ;)'
                ])),
            ])
        )
    )
)
```

### CODE REVIEW

**Comment by kitsunet** [comment]
Do not merge yet, probably want to refactor some of the code in the DTOs.  I think we should make the "collection of references for one reference property" an explicit DTO, as the code like this feels a bit too messy for my taste...

**Comment by kitsunet** [comment]
On the upside the somewhat weird and unfinished reference snapshots can be removed with this.

**Comment by kitsunet** [comment]
And forgot to adapt legacy migration tests (which rely on specific event payloads, see below, before I adjust those we should decide which way to go with b/c)...

**Comment by kitsunet** [comment]
oh and I should mention that at the moment this is not forgiving to the existing events. We _could_ do this in the respective fromArray constructors of the DTOs and just "upcast" from the old structure on the fly without much problem, or add a migration for the events.

File: Neos.ContentRepository.Core/Classes/Feature/Common/NodeReferencingInternals.php
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/Common/NodeReferencingInternals.php]:
```suggestion
        foreach ($references as $reference) {
```

isnt it part of the `IteratorAggregate` thing that one doesnt have to do it?

File: Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/NodeReferenceNameToEmpty.php
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/NodeReferenceNameToEmpty.php]: hu, i thought we wouldnt need a `*Unset` class like this as only the references will be set that are part of the `SetNodeReferences` command and passing an empty array will unset all references?

Also i had to read twice before i understood `ToEmpty` literally ... im not super happy with it :D

File: Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/NodeReferenceToWrite.php
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/NodeReferenceToWrite.php]: i would rather like to inline this code to the caller instead of here?
or a generic `toArray`? Idk or make it internal :D

File: Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/NodeReferencesToWrite.php
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/NodeReferencesToWrite.php]: im not sure if we really like to start annotating the fromArray shape as its a lost battle in terms of time to invest :D But thanks either way ^^

File: Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/NodeReferenceToWrite.php
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/NodeReferenceToWrite.php:34]: thanks for improving the api for setting references ill look into it how this new way is to be used instead and what is to be migrated for Neos 9 users...

While fine tuning the references api id like to also note that it often came to my mind that it didnt seem to have gotten the same love as setting properties. I think both setting references and properties should feel similar and not be totally a different thing.

**Review by mhsdesign — COMMENTED** [review]
Thanks for tackling this, some early thoughts before ill play with this ;)

File: Neos.ContentRepository.Core/Classes/Feature/Common/NodeReferencingInternals.php
Comment (kitsunet) [reply Neos.ContentRepository.Core/Classes/Feature/Common/NodeReferencingInternals.php]: right, right, I messed with this a bit, there are probably more artifacts like this in there.

File: Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/NodeReferenceNameToEmpty.php
Comment (kitsunet) [reply Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/NodeReferenceNameToEmpty.php]: Yep naming TBD, I am not super sold on this either, AND if we introduced the "in between" DTO as said above we wouldn't need this but would just say "NodeReferencesForProperty" and have that be empty for the specific reference name.

File: Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/NodeReferenceToWrite.php
Comment (kitsunet) [reply Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/NodeReferenceToWrite.php]: yeah internal I guess... But htis might again become easier with the middle DTO.

**Comment by kitsunet** [comment]
Screenshot 2024-06-18 at 08 47 52

I think this would be much nicer code wise (obviously duplicated for Serialized...), bit annoying to use 3 objects for this, but I think it will make the code much better to reason with and empty is then really just the new "middle DTO" with createEmpty()

**Comment by mhsdesign** [comment]
I remember we discussed this some time ago so i was wondering where it goes from now? I think we discussed for an alternative way than to use a "hacky" NodeReferenceNameToEmpty?

File: Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/NodeReferenceNameToEmpty.php
Comment (kitsunet) [reply Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/NodeReferenceNameToEmpty.php]: this has been removed

File: Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/NodeReferenceToWrite.php
Comment (kitsunet) [reply Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/NodeReferenceToWrite.php]: This has been simplified

File: Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php
Comment (mhsdesign) [inline Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php]: we should filter here by type `NodeReferencesWereSet` so its cheaper :)

File: Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/05-NodeReferencing/04-SetNodeReferences_PropertyScopes.feature
Comment (mhsdesign) [inline Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/05-NodeReferencing/04-SetNodeReferences_PropertyScopes.feature:162]:
```suggestion
  Scenario: Set node references in single command
```

File: Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/05-NodeReferencing/04-SetNodeReferences_PropertyScopes.feature
Comment (mhsdesign) [inline Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/05-NodeReferencing/04-SetNodeReferences_PropertyScopes.feature]:
```suggestion
  Scenario: Set node references in separate commands
```

File: Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/NodeCopying/CopyNode_NoDimensions.feature
Comment (mhsdesign) [inline Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/NodeCopying/CopyNode_NoDimensions.feature:71]: nice :D and does it also work across dimensions? see https://github.com/neos/neos-development-collection/issues/4614 ...it should work right?

File: Neos.ContentRepository.Core/Classes/Feature/Common/NodeReferencingInternals.php
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/Common/NodeReferencingInternals.php]:
```suggestion
```

File: Neos.ContentRepository.Core/Classes/Feature/NodeCreation/Command/CreateNodeAggregateWithNode.php
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/NodeCreation/Command/CreateNodeAggregateWithNode.php]:
```suggestion
    /**
```

sneaky marker :D

File: Neos.ContentRepository.Core/Classes/Feature/NodeCreation/Command/CreateNodeAggregateWithNode.php
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/NodeCreation/Command/CreateNodeAggregateWithNode.php]: i think this should not be nullable and like `PropertyValuesToWrite` ... that should be easier to handle

File: Neos.ContentRepository.Core/Classes/Feature/NodeCreation/Command/CreateNodeAggregateWithNodeAndSerializedProperties.php
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/NodeCreation/Command/CreateNodeAggregateWithNodeAndSerializedProperties.php]: i think it should not be nullable, but we can just use an empty instance here

File: Neos.ContentRepository.Core/Classes/Feature/NodeDuplication/Dto/NodeReferenceSnapshot.php
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/NodeDuplication/Dto/NodeReferenceSnapshot.php:1]: nice

**Review by mhsdesign — COMMENTED** [review]
ill re review later:)

File: Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/SerializedNodeReferencesForName.php
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/SerializedNodeReferencesForName.php]:
```suggestion
final readonly class SerializedNodeReferencesForName
```

File: Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/SerializedNodeReference.php
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/SerializedNodeReference.php]:
```suggestion
 * @api used in commands and events {@see SerializedNodeReferences}
```

File: Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php
Comment (mhsdesign) [inline Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php:1]: lecker heute wird wieder array Suppe gekocht :D

File: Neos.ContentRepository.LegacyNodeMigration/Classes/NodeDataToEventsProcessor.php
Comment (mhsdesign) [inline Neos.ContentRepository.LegacyNodeMigration/Classes/NodeDataToEventsProcessor.php:485]: todo

Comment (kitsunet) [reply Neos.ContentRepository.LegacyNodeMigration/Classes/NodeDataToEventsProcessor.php:485]: Nod, but I really do not know. I can't fix this and it wasn't the case before so 🤷 not worse than before...

File: Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php
Comment (mhsdesign) [reply Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php]: fyi we cannot use `NodeReferencesWereSet::class` here ... it HAS to be `"NodeReferencesWereSet"` :)

File: Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php
Comment (kitsunet) [inline Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php:264]: Somehow this API seems not really nice... But yeah I misread the other place we use it

File: Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php
Comment (kitsunet) [inline Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php:268]: we could _probably_ even skip the check altogether now with the filter?

Comment (mhsdesign) [reply Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php:268]: jup but when writing these kind of things i prefer to have this as sanity check :D helps sometimes ^^

File: Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php
Comment (mhsdesign) [reply Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php:264]: haha jup @bwaidelich and me already "complained" about that too :D

File: Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php
Comment (bwaidelich) [inline Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php]: Couldn't this be sth like
```
DELETE FROM $table WHERE name IN (names)
```

File: Neos.ContentGraph.PostgreSQLAdapter/src/Domain/Projection/Feature/NodeReferencing.php
Comment (bwaidelich) [inline Neos.ContentGraph.PostgreSQLAdapter/src/Domain/Projection/Feature/NodeReferencing.php:72]: Same here, can't we turn this into two atomic queries?

File: Neos.ContentRepository.Core/Classes/Feature/NodeDuplication/Dto/NodeSubtreeSnapshot.php
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Feature/NodeDuplication/Dto/NodeSubtreeSnapshot.php]: nitpick: in other places we use US english (serialise vs serialize). Furthermore @mhsdesign suggests to rename this:
```suggestion
    private static function serializeProjectedReferences(References $references): SerializedNodeReferences
```

File: Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/NodeReferencesForName.php
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/NodeReferencesForName.php:25]: some more docs about what this is about would be nice :)

File: Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/NodeReferencing.php
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/NodeReferencing.php]: Is there a specific reason why we _don't_?
The current state would be like publishing multiple `NodePropertiesWereSet` events for each property.

Couldn't we just build up the `$referencesByName` array and then build the event outside of the loop?

File: Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php
Comment (bwaidelich) [reply Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php:264]: yes, can be improved in the eventstore package to allow for simple types, but that's something for another day

**Review by bwaidelich — COMMENTED** [review]
Great work.. and a lot of it, yikes.
Some rather minor issues

**Comment by mhsdesign** [comment]
and the test has to be adjusted as it fails now earlier:

```
001 Scenario: Node reference cannot hold multiple targets to the same node                                          # Features/05-NodeReferencing/01-SetNodeReferences_ConstraintChecks.feature:190
      Then the last command should have thrown an exception of type "InvalidArgumentException" with code 1700150910 # Features/05-NodeReferencing/01-SetNodeReferences_ConstraintChecks.feature:195
        Expected exception code 1700150910, got exception code 1730365958 instead; Message: Duplicate entry in references to write. Target "anthony-destinode" already exists in collection.
        Failed asserting that 1730365958 is identical to 1700150910.
```

**Comment by mhsdesign** [comment]
Okay i fixed the tests again and added some documentation ... i also got rid of the duplicate `NodeReferencesToWrite::fromReferences` and its now just `create` and also the `fromNameAnd` parts are out for the `NodeReferencesForName` as well its guaranteed to stick to a name and thus makes it a little shorter and hopefully easy to use.

```php
SetNodeReferences::create(
    $node->workspaceName,
    $node->aggregateId,
    $node->originDimensionSpacePoint,
    NodeReferencesToWrite::create(
        // sets the 3 nodes as references
        NodeReferencesForName::fromTargets(
            ReferenceName::fromString('first-reference'),
            NodeAggregateIds::fromArray(['node-aggregate-id-1', 'node-aggregate-id-2', 'node-aggregate-id-3'])
        ),
        // unset previously set references (as we do NOT merge the values)
        NodeReferencesForName::createEmpty(
            ReferenceName::fromString('second-reference')
        ),
        // create a simple node reference and additionally a reference with an additional property
        NodeReferencesForName::fromReferences(
            ReferenceName::fromString('third-reference'),
            [
                NodeReferenceToWrite::fromTarget(NodeAggregateId::fromString('node-aggregate-id-6')),
                NodeReferenceToWrite::fromTargetAndProperties(NodeAggregateId::fromString('node-aggregate-id-5'), PropertyValuesToWrite::fromArray([
                    'propertyOnReference' => 'Hi im living on the edge ;)'
                ])),
            ])
    )
)
```

File: Neos.ContentGraph.PostgreSQLAdapter/src/Domain/Projection/Feature/NodeReferencing.php
Comment (mhsdesign) [reply Neos.ContentGraph.PostgreSQLAdapter/src/Domain/Projection/Feature/NodeReferencing.php:72]: for me the pg adapter is out of scope and there are no tests ... but i fixed it for doctrine

File: Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/NodeReferencing.php
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/NodeReferencing.php]: Adjusted:)

**Review by mhsdesign — APPROVED** [review]
Okay i further made the optimisations basti suggested. Now its truly finished id say <3

Thank you so much for your effort!

**Pull request merged**
